src/coroutine.rs

```diff
@@ -0,0 +1,230 @@
+//! Python coroutine implementation, used notably when wrapping `async fn`
+//! with `#[pyfunction]`/`#[pymethods]`.
+use crate::exceptions::{PyRuntimeError, PyStopIteration};
+use crate::pyclass::IterNextOutput;
+use crate::sync::GILOnceCell;
+use crate::types::{PyCFunction, PyIterator};
+use crate::{intern, wrap_pyfunction, IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use pyo3_macros::{pyclass, pyfunction, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Coroutine `throw` callback, see [`Coroutine::new`] about its use.
+pub type ThrowCallback = Box<dyn FnMut(&PyAny) + Send>;
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    throw_callback: Option<ThrowCallback>,
+    waker: Option<Arc<AsyncioWaker>>,
+}
+
+impl Coroutine {
+    /// Wrap a boxed future in to a Python coroutine.
+    ///
+    /// Coroutine `send` polls the wrapped future, ignoring the value passed
+    /// (should always be `None` anyway).
+    ///
+    /// If `throw_callback` has been provided, coroutine `throw` calls it
+    /// with the exception passed, and poll the wrapped future.
+    /// Otherwise, the exception is reraised and the future dropped.
+    pub fn new(
+        future: Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>,
+        throw_callback: Option<ThrowCallback>,
```

Comment:
I think given that our first pass at the `async` syntax doesn't have a way to deal with `throw_callback`, does it make sense to break this out into a separate PR? I am not mandating it, just wondering if we want to keep this API private for now.

(And if we do keep this API private, do we even need to make the `Coroutine` type public? If it's private, that might make it easier for us to iterate on functionality in the short term.)

Reply:
E.g. do we want a `pyo3(on_cancel = ...)` option for this, or will it always require the user manually constructing the Coroutine? It feels like an additional feature over and above the basics, and more for me to get my head around 😵‍💫

Reply:
You convinced me, and I've removed `throw_callback` (I can revert it back ofc), but I came with an new way of handling cancellation, see https://github.com/PyO3/pyo3/pull/3540#issuecomment-1784128394.

However, I think it's necessary to keep `Coroutine` public, especially for capture issue – you may not want to capture all the parameters of your function into your future, so you may want to have a sync function returning a `Coroutine` instead (the same way you would return `impl Future<Output=...>` instead of using an async function).
It also allows to have synchronous code executed before returning the coroutine.